### DISCUSSION TITLE
Breaking Customized change in 3.x

### DISCUSSION THREAD

**Discussion by PavelVanecek** [recharts/recharts#4437] [Polls]
## Context

Okay so in 3.x we're changing how Recharts works internally and we're mostly able to do that without breaking the API. With one exception and that is [`Customized`](https://recharts.org/en-US/api/Customized). This component exists as a workaround to the fact that Recharts will reject any custom components (https://github.com/recharts/recharts/issues/2788). It also provides (and this is not documented) [access to full internal state of Recharts](https://github.com/recharts/recharts/blob/f4f34716c288c4538a81b6a20ed1305dbbe725ea/src/chart/generateCategoricalChart.tsx#L1847).

## Why the breaking changes?

Now for the sad part: the full access to internal state means we can't change that state without breaking the API. And that state happens to be in dire need of a change. So I really want to change that state which means breaking changes to Customized API.

## Good news

And now the happy part: when 3.x is released, Recharts will allow rendering any custom components, without the need of the `Customized` wrapper. So that's good and one of the reasons for its existence will go away.

## What I want to do in 3.x

To make up for the lost access to internal state, we are going to introduce a set of React hooks. So instead of reading (undocumented and untyped) state from Customized component, users of the library will be able to use hooks to get access to internal and computed state of Recharts.

This will be documented, with TypeScript types out of the box, and examples.

## Preview (the 3.x version may change)

We don't have the API locked in yet but it might look something like this:

```jsx
import { BarChart, useChartViewPort } from 'recharts'

function MyCustomComponent() {
  const viewPort = useChartViewPort()
  return <rect x={viewPort.x} y={viewPort.y} />
}

function MyChart() {
  return (
    <BarChart>
      <MyCustomComponent />
    </BarChart>
  )
}
```

## That was the longest intro ever, what's the question?

**Comment by ckifer** [comment]
As long as the most important data is accessible via hooks, I say lets remove it. The most important thing (imo) is to be able to translate two values X and Y (i.e.  "Page A", 2000 or 2000, 2000) into chart coordinates. That doesn't even exist in `Customized` today and would be incredibly powerful

**Comment by PavelVanecek** [comment]
Thanks everyone for your comments and votes! In conclusion: 3.0 will ship without `Customized`: https://github.com/recharts/recharts/issues/4551

**Comment by NarendraKuruva** [comment]
@ckifer @PavelVanecek 

When is Version 3.0 set to be released??

**Reply by PavelVanecek** [reply]
We don't have a date set yet. After all tasks from here are done: https://github.com/orgs/recharts/projects/2

**Reply by NarendraKuruva** [reply]
Any expected/tentative time?

To plan our next things accordingly
@PavelVanecek

**Reply by ckifer** [reply]
There can't be an expected date until we're almost there without being very incorrect, sorry.

**Reply by PavelVanecek** [reply]
@NarendraKuruva 3.0.0 alpha version is now available: https://github.com/recharts/recharts/discussions/5444
